code/controllers/subsystem/demo.dm

```diff
@@ -0,0 +1,425 @@
+				T.demo_last_appearance = T.appearance
+				var/this_appearance
+				// space turfs are difficult to RLE otherwise, because they all
+				// have different appearances despite being the same thing.
+				if(T.type == /turf/open/space || T.type == /turf/open/space/basic)
+					this_appearance = "s" // save the bytes
+				else if(istype(T, /turf/open/space/transit))
+					this_appearance = "t[T.dir]"
+				else
+					this_appearance = T.appearance
+				if(this_appearance == last_appearance)
+					rle_count++
+				else
+					if(rle_count > 1)
+						row_list += rle_count
+					rle_count = 1
+					if(istext(this_appearance))
+						row_list += this_appearance
+					else
+						// do a diff with the previous turf to save those bytes
+						row_list += encode_appearance(this_appearance, istext(last_appearance) ? null : last_appearance)
+				last_appearance = this_appearance
+		if(rle_count > 1)
+			row_list += rle_count
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, jointext(row_list, ",") + "\n")
+	CHECK_TICK
+	// then do objects
+	world.log << "Writing objects"
+	marked_new.len = 0
+	marked_dirty.len = 0
+	for(var/z in 1 to world.maxz)
+		var/spacing = 0
+		var/row_list = list()
+		for(var/y in 1 to world.maxy)
+			for(var/x in 1 to world.maxx)
+				var/turf/T = locate(x,y,z)
+				var/list/turf_list = list()
+				for(var/C in T.contents)
+					var/atom/movable/as_movable = C
+					if(as_movable.loc != T)
+						continue
+					if(isobj(C) || ismob(C))
+						turf_list += encode_init_obj(C)
+				if(turf_list.len)
+					if(spacing)
+						row_list += spacing
+						spacing = 0
+					row_list += turf_list
+				spacing++
+			CHECK_TICK // This is a bit risky because something might change but meh, its not a big deal.
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, jointext(row_list, ",") + "\n")
+
+	// track objects that exist in nullspace
+	var/nullspace_list = list()
+	for(var/atom/movable/M in world)
+		if(M.loc != null) continue
+		if(!isobj(M) && !ismob(M))
+			continue
+		nullspace_list += encode_init_obj(M)
+		CHECK_TICK
+	WRITE_LOG_NO_FORMAT(GLOB.demo_log, jointext(nullspace_list, ",") + "\n")
+
+	for(var/line in pre_init_lines)
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, "[line]\n")
+
+	return ..()
+
+/datum/controller/subsystem/demo/fire()
+	if(!src.marked_new.len && !src.marked_dirty.len && !src.marked_turfs.len && !src.del_list.len)
+		return // nothing to do
+
+	last_queued = src.marked_new.len + src.marked_dirty.len + src.marked_turfs.len
+	last_completed = 0
+
+	write_time()
+	if(src.del_list.len)
+		var/s = "del [jointext(src.del_list, ",")]\n" // if I don't do it like this I get "incorrect number of macro arguments" because byond is stupid and sucks
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, s)
+	src.del_list.len = 0
+
+	var/canceled = FALSE
+
+	var/list/marked_dirty = src.marked_dirty
+	var/list/dirty_updates = list()
+	while(marked_dirty.len)
+		last_completed++
+		var/atom/movable/M = marked_dirty[marked_dirty.len]
+		marked_dirty.len--
```

Comment:
This should be done once at the end of iteration and once `MC_TICK_CHECK` succeeds instead of for every item

Reply:
I think SSair does something like that

Reply:
This is how SSair does it as well, not sure what you're looking for.